differs/rpm_diff.go

```diff
@@ -364,3 +366,115 @@ func unlock() error {
+	analysis, err := singleVersionLayerAnalysis(image, a)
+	return analysis, err
+}
+
+// getPackages returns an array of maps of installed rpm packages on each layer
+func (a RPMLayerAnalyzer) getPackages(image pkgutil.Image) ([]map[string]util.PackageInfo, error) {
+	path := image.FSPath
+	var packages []map[string]util.PackageInfo
+	if _, err := os.Stat(path); err != nil {
+		// invalid image directory path
+		return packages, err
+	}
+
+	// try to find the rpm binary in bin/ or usr/bin/
+	rpmBinary := filepath.Join(path, "bin/rpm")
+	if _, err := os.Stat(rpmBinary); err != nil {
+		rpmBinary = filepath.Join(path, "usr/bin/rpm")
+		if _, err = os.Stat(rpmBinary); err != nil {
+			logrus.Errorf("Could not detect RPM binary in unpacked image %s", image.Source)
+			return packages, nil
+		}
+	}
+
+	packages, err := rpmDataFromLayerFS(image)
+	if err != nil {
+		logrus.Info("Running RPM binary from image in a container")
+		return rpmDataFromLayeredContainers(image.Image)
+	}
+	return packages, err
+}
+
+// rpmDataFromLayerFS runs a local rpm binary, if any, to query the layer
+// rpmdb and returns an array of maps of installed packages.
+func rpmDataFromLayerFS(image pkgutil.Image) ([]map[string]util.PackageInfo, error) {
+	var packages []map[string]util.PackageInfo
+	// Check there is an executable rpm tool in host
+	if err := exec.Command("rpm", "--version").Run(); err != nil {
+		logrus.Warn("No RPM binary in host")
+		return packages, err
+	}
+	dbPath, err := rpmDBPath(image.FSPath)
+	if err != nil {
+		logrus.Warnf("Couldn't find RPM database: %s", err.Error())
+		return packages, err
+	}
+	for _, layer := range image.Layers {
+		layerPackages := make(map[string]util.PackageInfo)
+		//query only layers that include the rpm database
+		if _, err := os.Stat(filepath.Join(layer.FSPath, dbPath)); err == nil {
+			cmdArgs := append([]string{"--root", layer.FSPath, "--dbpath", dbPath}, rpmCmd[1:]...)
+			out, err := exec.Command(rpmCmd[0], cmdArgs...).Output()
+			if err != nil {
+				logrus.Warnf("RPM call failed: %s", err.Error())
+				return packages, err
+			}
+			layerPackages, err = parsePackageData(strings.Split(string(out), "\n"))
+			if err != nil {
+				return packages, err
+			}
+		}
+		packages = append(packages, layerPackages)
+	}
+
+	return packages, nil
+}
+
+// rpmDataFromLayeredContainers runs a tmp image in a container for each layer,
+// queries the data of installed rpm packages and returns an array of maps of
+// packages.
+func rpmDataFromLayeredContainers(image v1.Image) ([]map[string]util.PackageInfo, error) {
+	var packages []map[string]util.PackageInfo
+	tmpImage, err := random.Image(0, 0)
+	if err != nil {
+		return packages, err
+	}
+	layers, err := image.Layers()
+	if err != nil {
+		return packages, err
+	}
+	// Append layers one by one to an empty image and query rpm
+	// database on each iteration
+	for _, layer := range layers {
+		tmpImage, err = mutate.AppendLayers(tmpImage, layer)
```

Comment:
We're trying to retrieve a map of packages installed in each layer here, right? I'm wondering if we actually want to be appending each layer onto the previous layers before running `rpmDataFromContainer()`, and instead create a single-layer image for each layer. Won't we be getting package information for the **combination of all layers** before the one we're actually examining here?

e.g. if I have layer `A` with package `foo==1.0.0`, and layer `B` with package `bar==1.0.0`, my maps here will look like 
`A`: {
  `foo`: `1.0.0`
},

`B`: {
  `foo`: `1.0.0`, <-- don't want this here right?
  `bar`: `1.0.0`,
}

To get around this you could just create a new `random.Image` for each layer, append it, and then make the `rpmDataFromContainer()` call. Does this make sense?

Reply:
> To get around this you could just create a new random.Image for each layer, append it, and then make the rpmDataFromContainer() call. Does this make sense?

This is the same situation we have in APT packages, the package database reports all the installed packages, thus in each layer we will get always the list of all packages included in current and previous layers. To work around this there is the method singleVersionLayerAnalysis  https://github.com/GoogleContainerTools/container-diff/blob/c2559537136c3d22b57f5a78f46c92064c93bfbc/differs/package_differs.go#L118-L141
 which basically performs a diff between layers.

Also I am appending the layers one by one because in this case the `rpm` binary used to parse the database is the one included inside the image, thus at least the layer including this binary (most probably the base layer) should be also part of the container to run. In order to make things simpler without having to include some kind of logic to guess which layers should be appended and which aren't needed I opted to simply stack them one by one. Note that this is the fallback in case the host does not have the RPM tool.

Reply:
Ahh right, ok I think I follow that. Thanks for the explanation!